src/fsm.c

```diff
@@ -435,27 +597,81 @@ static int fsm__snapshot(struct raft_fsm *fsm,
 	QUEUE__FOREACH(head, &f->registry->dbs)
 	{
 		db = QUEUE__DATA(head, struct db, queue);
-		rv = encodeDatabase(db, &(*bufs)[i]);
+		/* database_header + num_pages + wal */
+		unsigned n = 1 + dbNumPages(db) + 1;
+		rv = encodeDatabase(db, &(*bufs)[i], n);
 		if (rv != 0) {
 			goto err_after_encode_header;
 		}
-		i += 2;
+		i += n;
 	}
 
+	assert(i == *n_bufs);
 	return 0;
 
 err_after_encode_header:
-	do {
-		raft_free((*bufs)[i].base);
-		i--;
-	} while (i > 0);
+	freeSnapshotBufs(f, *bufs, i);
 err_after_bufs_alloc:
-	raft_free(*bufs);
+	sqlite3_free(*bufs);
 err:
+	QUEUE__FOREACH(head, &f->registry->dbs)
+	{
+		db = QUEUE__DATA(head, struct db, queue);
+		databaseReadUnlock(db);
+	}
 	assert(rv != 0);
 	return rv;
 }
 
+static int fsm__snapshot_finalize(struct raft_fsm *fsm,
+				  struct raft_buffer *bufs[],
+				  unsigned *n_bufs)
+{
+	struct fsm *f = fsm->data;
+	queue *head;
+	struct db *db;
+	unsigned n_db;
+	struct snapshotHeader header;
+	int rv;
+
+	if (bufs == NULL) {
+		return 0;
+	}
+
+	/* Decode the header to determine the number of databases. */
+	struct cursor cursor = {(*bufs)[0].base, (*bufs)[0].len};
+	rv = snapshotHeader__decode(&cursor, &header);
+	if (rv != 0) {
+		tracef("decode failed %d", rv);
+		return -1;
+	}
+	if (header.format != SNAPSHOT_FORMAT) {
+		tracef("bad format");
+		return -1;
+	}
+
+	/* Free allocated buffers */
+	freeSnapshotBufs(f, *bufs, *n_bufs);
+	sqlite3_free(*bufs);
+	*bufs = NULL;
+	*n_bufs = 0;
+
+	/* Unlock all databases that were locked for the snapshot, this is safe
+	 * because DB's are only ever added at the back of the queue. */
```

Comment:
Are there are cases where DBs are removed? I guess currently not.

Reply:
Indeed, currently not.